### PR TITLE
feat: add TierTemplate CRD

### PR DESCRIPTION
**Pull request by xcoulon**
## Description
The TierTemplate custom resources carry an OpenShift template
along with a type and revision so we know to which namespace
(or cluster-wide resources) the template applies, and what
is its revision.
Each time a template is updated, a new custom resource is
created and the revision is the commit hash of the corresponding
file in https://github.com/codeready-toolchain/host-operator/tree/master/deploy/templates/nstemplatetiers

## Checks
1. Have you run `make generate` target? **yes**

2. Does `make generate` change anything in other projects (host-operator, member-operator, toolchain-common)? **yes** 

3. In case other projects are changed, please provides PR links.
    - host-operator: https://github.com/codeready-toolchain/host-operator/pull/187
    - member-operator: https://github.com/codeready-toolchain/member-operator/pull/146
 

Fixes https://issues.redhat.com/browse/CRT-499

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: pkg/apis/addtoscheme_toolchain_v1alpha1.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package apis
+
+import (
+	"github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+)
+
+func init() {
+	// Register the types with the Scheme so the components can map objects to GroupVersionKinds and back
+	AddToSchemes = append(AddToSchemes, v1alpha1.SchemeBuilder.AddToScheme)
+}
```
Comment (MatousJobanek) [inline pkg/apis/addtoscheme_toolchain_v1alpha1.go:10]: we don't need this file - we already add the types to the scheme here: https://github.com/codeready-toolchain/api/blob/master/pkg/apis/apis.go

File: pkg/apis/toolchain/group.go [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// Package toolchain contains toolchain API versions.
+//
+// This file ensures Go source parsers acknowledge the toolchain package
+// and any child packages. It can be removed if any other Go source files are
+// added to this package.
+package toolchain
```
Comment (MatousJobanek) [inline pkg/apis/toolchain/group.go:6]: Do we need this file? Technically, it's not a problem having it here,  I'm just curious where it came from...

File: pkg/apis/toolchain/v1alpha1/nstemplateset_types.go [hunk 3]
```diff
@@ -59,6 +59,10 @@ type NSTemplateSetNamespace struct {
 	// Optional field. Used to specify a custom template
 	// +optional
 	Template string `json:"template,omitempty"`
+
+	// TemplateRef the name of the TemplateTier resource on the host cluster and which contains the template to use
```
Comment (MatousJobanek) [inline pkg/apis/toolchain/v1alpha1/nstemplateset_types.go:63]: I would use slightly different wording - it's (after all) used in the description withing the CRD. 
In addition, you were apparently influenced by my mistake with the name - it's not `TemplateTier` but `TierTemplate` :-) 
how about having something like this:
```suggestion
	// The name of the TierTemplate resource that exists in the host cluster and that contains the template to be used
```

File: pkg/apis/toolchain/v1alpha1/tiertemplate_types.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+// +k8s:openapi-gen=true
+type TierTemplateSpec struct {
+	
+	// The type of the template. For example: "code", "dev" or "stage"
+	Type string `json:"type"`
+
+	// The revision of the corresponding template
+	Revision string `json:"revision"`
+
+	// Template the OpenShift Template to be used to provision a user's namespace or cluster-wide resources
```
Comment (MatousJobanek) [inline pkg/apis/toolchain/v1alpha1/tiertemplate_types.go:18]: I would use the 
```suggestion
	// Template contains an OpenShift Template to be used to provision either a user's namespace or cluster-wide resources
```

File: pkg/apis/toolchain/v1alpha1/tiertemplate_types.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import (
+	templatev1 "github.com/openshift/api/template/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// TierTemplateSpec defines the desired state of TierTemplate
+// +k8s:openapi-gen=true
+type TierTemplateSpec struct {
+	
+	// The type of the template. For example: "code", "dev" or "stage"
```
Comment (MatousJobanek) [inline pkg/apis/toolchain/v1alpha1/tiertemplate_types.go:12]: I know that it lists only the examples, but I would add also the "cluster" type
```suggestion
	// The type of the template. For example: "code", "dev", "stage" or "cluster"
```

File: pkg/apis/toolchain/v1alpha1/tiertemplate_types.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	Revision string `json:"revision"`
+
+	// Template the OpenShift Template to be used to provision a user's namespace or cluster-wide resources
+	Template templatev1.Template `json:"template"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// TierTemplate is the Schema for the tiertemplates API
+// +kubebuilder:resource:path=tiertemplates,scope=Namespaced
```
Comment (MatousJobanek) [inline pkg/apis/toolchain/v1alpha1/tiertemplate_types.go:25]: What about adding `type` and `revision` columns?
```suggestion
// +kubebuilder:resource:path=tiertemplates,scope=Namespaced
// +kubebuilder:printcolumn:name="Type",type="string",JSONPath=`.spec.type`
// +kubebuilder:printcolumn:name="Revision",type="string",JSONPath=`.spec.revision`
```

File: pkg/apis/toolchain/group.go (hunk 2 above)
Comment (xcoulon) [reply pkg/apis/toolchain/group.go:6]: my apologies, I did not realize this file was part of the PR. I don't even know how it was generated, and I'll remove it.

Comment (xcoulon) [reply pkg/apis/toolchain/group.go]: done in 0010f97

File: pkg/apis/addtoscheme_toolchain_v1alpha1.go (hunk 1 above)
Comment (xcoulon) [reply pkg/apis/addtoscheme_toolchain_v1alpha1.go:10]: good point! Removed in c462ff0

**Comment by xcoulon** [comment]
Note: codeready-toolchain/host-operator#187 and codeready-toolchain/member-operator#146 have been updated accordingly

**Comment by xcoulon** [comment]
> I'm not sure why, but you again reverted the changed comment at `TemplateRef` variable to the original value - correct comment should be:
> 
> ```
> // The name of the TierTemplate resource that exists in the host cluster and that contains the template to be used
> ```
> 
> could you please change all 4 occurrences?
> In addition, you again committed the `pkg/apis/addtoscheme_toolchain_v1alpha1.go` file - please remove it.

sorry about that, I read your previous comment in the wrong order 🤦‍♂️

**Comment by xcoulon** [comment]
> > I'm not sure why, but you again reverted the changed comment at `TemplateRef` variable to the original value - correct comment should be:
> > ```
> > // The name of the TierTemplate resource that exists in the host cluster and that contains the template to be used
> > ```
> > 
> > 
> > could you please change all 4 occurrences?
> > In addition, you again committed the `pkg/apis/addtoscheme_toolchain_v1alpha1.go` file - please remove it.
> 
> sorry about that, I read your previous comment in the wrong order 🤦‍♂️

done in 0904dfb

**Pull request merged**